### ISSUE TITLE
Resource Timing Metrics

### ISSUE THREAD

**Issue by JohnRiv** [Comcast/Surf-N-Perf#41]
Here's what I want to build:
- a "Resource Timing" add-on with the following functions: (note for all functions below, if the browser does not support resource timing (i.e. window.performance.getEntriesByType does not exist), then every function should return `null`)
  - **getOrigins(options)** - returns an array of all the origins (e.g. "https://github.com") returned by window.performance.getEntriesByType("resource") with no duplicates. The options passed to it is an object that includes either a "whitelist" key with a single string or an array of origins as the value that will be used to return only those items that are in the list of origins or a "blacklist" key with a single string or an array of origins as the value that will be used to filter out those origins
    - IE may include other non-fetched resources, such as about:blank and javascript: URLs in the ResourceTiming data, so also filter out non-http: and https: protocols.
  - **getResourcesFromOrigin(origin)** - returns an array of all the resources ~names~ from the specified origin
  - **getResourceNamesFromOrigin(origin)** - returns an array of all the resource names from the specified origin
  - ~~**isBufferFull()** - returns `false` (if resource timing is supported), unless the `performance.onresourcetimingbufferfull` or `performance.onwebkitresourcetimingbufferfull` event fired, then returns `true`~~

For all of these functions below, the "name" argument is full request, e.g. "http://johnriv.github.io/Surf-N-Perf/docs/scripts/linenumber.js". If the argument starts with a "/", then use the page's current origin. For example:
surfnperfRT.getResource("/Surf-N-Perf/docs/scripts/linenumber.js") would return window.performance.getEntriesByName("http://johnriv.github.io/Surf-N-Perf/docs/scripts/linenumber.js", "resource")[0] if it was called from http://johnriv.github.io/Surf-N-Perf/ 
- **duration(name, eventA, eventB, options)** - returns the diff between 2 resource timing marks for a single resource (`name`). Takes optional 4th param which can be {decimalPlaces:NUMBER}, just like Surf-N-Perf's. If eventA or eventB's value is `0`, return `false`. Returns `undefined` if the resource is not found
- **getResource(name)** - shorthand for window.performance.getEntriesByName(name, "resource")[0]
- **start(name, options)** - shorthand for window.performance.getEntriesByName(name, "resource")[0].startTime. Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found
- **end(name, options)** - shorthand for window.performance.getEntriesByName(name, "resource")[0].responseEnd. Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found
- **getFullRequestLoadTime(name, options)** - returns the 'duration' value for the specified resource from window.performance.getEntriesByName(name, "resource")[0].  Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found.
- **getNetworkTime(name, options)** - very similar to Surf-N-Perf's getNetworkTime in that it returns the duration between 'fetchStart' & 'connectEnd' for the resource.   Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found. Returns `false` if connectEnd is 0 (3rd-party without a `Timing-Allow-Origin: *` header)
- **getServerTime(name, options)** - very similar to Surf-N-Perf's getServerTime in that it returns the duration between 'requestStart' & 'responseEnd' for the resource.   Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found. Returns `false` if requestStart is 0 (3rd-party without a `Timing-Allow-Origin: *` header)
- **getBlockingTime(name,options)** - see http://nicj.net/resourcetiming-in-practice/ for logic.    Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found. Returns `false` if requestStart is 0 (3rd-party without a `Timing-Allow-Origin: *` header)
- **getNetworkDuration(name,options)** - see http://www.stevesouders.com/blog/2014/11/25/serious-confusion-with-resource-timing/ for logic.    Optional 2nd param for {decimalPlaces:NUMBER}. Returns `undefined` if the resource is not found. Returns `false` if requestStart is 0 (3rd-party without a `Timing-Allow-Origin: *` header)

And here's how it should be built:
- create a new file at /lib/resource-timing.js
  - add the same copyright notice to that as surfnperf.js
  - the Gruntfile.js will need to be updated so uglify creates a /resource-timing.js from /lib/resource-timing.js, jshint checks /lib/resource-timing.js and watch watches /lib/resource-timing.js
  - the package.json will need to be updated to include /lib/resource-timing.js & /resource-timing.js in the list of files
- The module definition should be defined as follows so that `surfnperf` is a dependency:

```
(function(root, factory) {
  if(typeof define === 'function' && define.amd) {
    // For Require.js
    define('surfnperf/resource-timing', ['surfnperf'], factory);
  } else if(typeof exports === 'object') {
    // For Browserify
    module.exports = factory(require('surfnperf'));
  } else {
    // Browser global if not using Require.js or Browserify
    root.surfnperfRT = factory(root.surfnperf);
  }
}(this, function(surfnperf) {
  // code goes here
}));
```
- All code should be 100% tested
- All methods should be fully documented in code with JSDoc comments
- Demo page should be updated to showcase new methods
- README.md & [JavaScript API](https://github.com/Comcast/Surf-N-Perf/wiki/JavaScript-API) should be updated explaining how to use them

As a reference, these links help explain the value of Resource Timing:
http://nicj.net/resourcetiming-in-practice/
http://www.stevesouders.com/blog/2014/11/25/serious-confusion-with-resource-timing/
https://paul.kinlan.me/detecting-injected-content/

**Comment by JohnRiv** [comment]
In the case of an `OPTIONS` request, there will be at least 2 entries, so should also have a way to easily request additional entries for the same URL

**Comment by JohnRiv** [comment]
Made a couple more tweaks:

- `getResourcesFromOrigin` can return either the full resource or a specified property
- Added `getResourceNamesFromOrigin` method, which returns an array of all the resource names from the specified origin
- `getResources` will return all resources if no argument is passed to it

**Comment by JohnRiv** [comment]
Closed via #61
